### DISCUSSION TITLE
Incremental sync using listener_add

### DISCUSSION THREAD

**Discussion by DanielViberg** [govim/govim#1211] [Q&A]
Hello,

I'm looking into implementing incremental sync for a vim9 lsp client https://github.com/yegappan/lsp. 
This solution seems to utilize listener_add api. If i read the code correct it seems like you're adding Lines to the a bufChangedChange struct and then build changeset from that and the other attributes. 

I assume I'm missing something since I'm wondering how you avoid this issue: 
https://github.com/vim/vim/issues/11385

**Comment by myitcv** [comment]
Hi @DanielViberg - it's been ages since I looked at this code, but from memory (and a quick glance) I do everything line-wise:

https://github.com/govim/govim/blob/669f6011070ba6b303ff3c1ea4c5243ec1e56878/cmd/govim/buffer_events.go#L119

**Reply by DanielViberg** [reply]
Thanks!

**Reply by DanielViberg** [reply]
I apologize in advance if my understanding of your go code is faulty. It tried implementing the buffer_events.go in pure vim9script: 

https://github.com/DanielViberg/lsp/blob/6d134bf51b83e40a4fd27e26c82df55e2f6bd654/autoload/lsp/lspserver.vim#L611-L688

On these lines: 
https://github.com/govim/govim/blob/669f6011070ba6b303ff3c1ea4c5243ec1e56878/cmd/govim/buffer_events.go#L171-L173
Are you modifying a cached version of the buffer that is later used by bufChanged?

**Reply by myitcv** [reply]
> I apologize in advance if my understanding of your go code is faulty.

It's not unlikely the code is unclear/incorrect :)

I might be misunderstanding your question, but I'll lay out the general setup:

* `*vimstate` is `govim`'s idea of state, based on messages received from Vim. 
* `(*vimstate).bufChanged` is called by Vim when a buffer has changed. It is therefore the job of this method to update `govim`'s notion of what a buffer holds.
* The code you have linked to is indeed mutating the state of the buffer known to `govim`

So generally speaking the code is modifying the state of the buffer known to the remote process that is `govim`. 

This is all required because `govim` is running in a separate process, hence needs to maintain some state for efficiency reasons. And `bufChanged` is essentially a key building block in that, handling deltas from Vim that express the change to buffer contents.

Does that help?

**Comment by DanielViberg** [comment]
Ok, then i believe I understand it correct. Thanks!

I've read your conversation with Bram: https://groups.google.com/g/vim_dev/c/L8jg6u47Dro/m/CbH3LNXdAQAJ
I'm still seeing issues with irregular behavior with plugins like auto-pair, or in general with o,u,r,dd operations. Currently I just send the hole buffer to the server if there are more than one change send from listener_add. 

Do you remember if this is/has been an issue in govim as well?
